modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/store/Store.es.js

```diff
@@ -0,0 +1,256 @@
+				throw new TypeError('`action` must be an object');
+			}
+
+			if (!action.type || typeof action.type !== 'string') {
+				throw new Error('`action.type` must be a non-empty string');
+			}
+
+			if (isDispatching) {
+				throw new Error('Illegal dispatch() call during dispatch');
+			}
+
+			try {
+				isDispatching = true;
+				state = currentReducer(state, action);
+			} finally {
+				isDispatching = false;
+			}
+
+			for (const subscriber of subscribers.values()) {
+				try {
+					subscriber();
+				} catch (error) {
+					console.error(error);
+				}
+			}
+		},
+
+		getState() {
+			if (isDispatching) {
+				throw new Error('Illegal getState() call during dispatch');
+			}
+
+			return state;
+		},
+
+		replaceReducer(nextReducer) {
+			currentReducer = nextReducer;
+
+			return store;
+		},
+
+		subscribe(subscriber) {
+			if (typeof subscriber !== 'function') {
+				throw new Error('`subscriber` must be a function');
+			}
+
+			if (isDispatching) {
+				throw new Error('Illegal subscribe() call during dispatch');
+			}
+
+			const nextSubscriberID = ++subscriberID;
+
+			subscribers.set(nextSubscriberID, subscriber);
+
+			return function unsubscribe() {
+				if (isDispatching) {
+					throw new Error(
+						'Illegal unsubscribe() call during dispatch'
+					);
+				}
+
+				subscribers.delete(nextSubscriberID);
+			};
+		},
+	};
+
+	return store;
+}
+
+/**
+ * Takes an object where the keys and values are namespaces and their
+ * corresponding reducer functions, and returns a single reducer function.
+ *
+ * See: https://redux.js.org/api/combinereducers
+ */
+export function combineReducers(reducers) {
+	const namespacesAndReducers = Object.entries(reducers);
+
+	return function combinedReducer(state, action) {
+		let hasChanged = false;
+
+		const nextState = {};
+
+		for ([namespace, reducer] of namespacesAndReducers) {
+			const previous = state[namespace];
+			const next = reducer(previous, action);
+
+			if (typeof next === 'undefined') {
+				throw new Error(`Reducer for ${namespace} returned undefined`);
+			}
+
+			hasChanged = hasChanged || previous !== next;
+
+			nextState[namespace] = next;
+		}
+
+		return hasChanged ? nextState : state;
+	};
+}
+
+export function defaultReducer(state, _action) {
+	return state;
+}
+
+/**
+ * Takes an array of reducers and returns a single reducer that runs them
+ * serially.
+ *
+ * This produces a "flat" state, unlike `combineReducers`, which produces a
+ * "nested/namespaced" state.
+ *
+ * See: https://github.com/redux-utilities/reduce-reducers
+ */
+export function reduceReducers(reducers) {
+	return function reducedReducer(state, action) {
+		return reducers.reduce((currentState, reducer, DELETE_ME) => {
+			const nextState = reducer(currentState, action);
+
+			if (typeof nextState === 'undefined') {
+				throw new Error(`Reducer for returned undefined`);
+			}
+
+			return nextState;
+		}, state);
+	};
+}
+
+const STORES = {};
+
+const SUB_QUEUE = {};
+
+export default {
```

Comment:
This probably doesn't even need to be exported but we might want to make this Store importable rather than just on the global namespace

Reply:
I like the idea of putting simple access to the store on the global namespace, and then anything "extra" would be separate. This way we still provide access to the store to anything on the page, but the tooling such as hooks or utils would require a more internal and direct import.